Author: Tyler Hicks <tyhicks@canonical.com>
Origin: backport, https://sourceforge.net/p/squashfs/mailman/message/36343213/
Description: Using macros, rather than raw octal values, better conveys the
 intent of masking off the setuid, setgid, and sticky bits.
Forwarded: yes
Bug-Ubuntu: https://launchpad.net/bugs/1779914

Index: squashfs-tools-4.3/squashfs-tools/unsquashfs.c
===================================================================
--- squashfs-tools-4.3.orig/squashfs-tools/unsquashfs.c
+++ squashfs-tools-4.3/squashfs-tools/unsquashfs.c
@@ -821,6 +821,8 @@ int set_attributes(char *pathname, int m
 	unsigned int xattr, unsigned int set_mode)
 {
 	struct utimbuf times = { time, time };
+	/* Mode bits that are only useful with root privileges */
+	mode_t root_mask = S_ISUID | S_ISGID | S_ISVTX;
 
 	if(utime(pathname, &times) == -1) {
 		ERROR("set_attributes: failed to set time on %s, because %s\n",
@@ -836,9 +838,9 @@ int set_attributes(char *pathname, int m
 			return FALSE;
 		}
 	} else
-		mode &= ~07000;
+		mode &= ~(root_mask);
 
-	if((set_mode || (mode & 07000)) && chmod(pathname, (mode_t) mode) == -1) {
+	if((set_mode || (mode & root_mask)) && chmod(pathname, (mode_t) mode) == -1) {
 		ERROR("set_attributes: failed to change mode %s, because %s\n",
 			pathname, strerror(errno));
 		return FALSE;
